### PR TITLE
Kill policy grace period

### PR DESCRIPTION
**Pull request by NDonatucci**

### CODE REVIEW

File: api/src/main/thrift/org/apache/aurora/gen/api.thrift [hunk 1]
```diff
@@ -238,6 +238,10 @@ union Resource {
   5: i64 numGpus
 }
 
+struct KillPolicy {
```
Comment (mauri) [inline api/src/main/thrift/org/apache/aurora/gen/api.thrift:242]: add doc (see others structs for examples)

File: api/src/main/thrift/org/apache/aurora/gen/api.thrift [hunk 2]
```diff
@@ -238,6 +238,10 @@ union Resource {
...
+  1: i64 gracePeriod   /** Seconds */
```
Comment (mauri) [inline api/src/main/thrift/org/apache/aurora/gen/api.thrift]: include seconds in the name, the API should be descriptive on what it's expecting (use something  like gracePeriodSecs). 
Add the doc above describing what's this.

File: api/src/main/thrift/org/apache/aurora/gen/api.thrift [hunk 3]
```diff
@@ -259,6 +263,9 @@ struct TaskConfig {
  18: optional bool production
  /** Task tier type. */
  30: optional string tier
+
+ 15: optional KillPolicy killPolicy
```
Comment (mauri) [inline api/src/main/thrift/org/apache/aurora/gen/api.thrift:269]: Add comment with doc (see other struct members)

File: src/main/java/org/apache/aurora/scheduler/mesos/MesosTaskFactory.java [hunk 4]
```diff
@@ -48,15 +48,18 @@
...
 import org.apache.mesos.Protos.ExecutorInfo;
+import org.apache.mesos.Protos.KillPolicy;
 import org.apache.mesos.Protos.Label;
 import org.apache.mesos.Protos.Labels;
 import org.apache.mesos.Protos.Offer;
 import org.apache.mesos.Protos.Port;
 import org.apache.mesos.Protos.Resource;
 import org.apache.mesos.Protos.TaskID;
 import org.apache.mesos.Protos.TaskInfo;
+
```
Comment (mauri) [inline src/main/java/org/apache/aurora/scheduler/mesos/MesosTaskFactory.java]: remove added empty line (let's keep the diff tidy)

File: src/main/java/org/apache/aurora/scheduler/mesos/MesosTaskFactory.java [hunk 5]
```diff
@@ -214,8 +217,19 @@ public TaskInfo createFrom(IAssignedTask task, Offer offer) throws SchedulerExce
           taskBuilder.setExecutor(execBuilder.build());
         } else {
           LOG.warn("Running Docker-based task without an executor.");
+
```
Comment (mauri) [inline src/main/java/org/apache/aurora/scheduler/mesos/MesosTaskFactory.java]: same here

File: src/main/java/org/apache/aurora/scheduler/mesos/MesosTaskFactory.java [hunk 6]
```diff
@@ -214,8 +217,19 @@ public TaskInfo createFrom(IAssignedTask task, Offer offer) throws SchedulerExce
...
           taskBuilder.setContainer(getDockerContainerInfo(dockerContainer, Optional.absent()))
-              .setCommand(CommandInfo.newBuilder().setShell(false));
+                     .setCommand(CommandInfo.newBuilder().setShell(false));
```
Comment (mauri) [inline src/main/java/org/apache/aurora/scheduler/mesos/MesosTaskFactory.java]: is the indentation ok? (compare with the removed line)

File: src/main/java/org/apache/aurora/scheduler/mesos/MesosTaskFactory.java [hunk 7]
```diff
@@ -214,8 +217,19 @@ public TaskInfo createFrom(IAssignedTask task, Offer offer) throws SchedulerExce
...
+
+          if (config.isSetKillPolicy()) {
+            DurationInfo.Builder durationBuilder;
+            durationBuilder = DurationInfo.newBuilder()
+                    .setNanoseconds(config.getKillPolicy().getGracePeriod() * 1000000000);
```
Comment (mauri) [inline src/main/java/org/apache/aurora/scheduler/mesos/MesosTaskFactory.java]: Use java's Duration.ofFoo().toBar() to do the conversion

**Review by mauri — CHANGES_REQUESTED** [review]
Great job!!

File: src/main/java/org/apache/aurora/scheduler/mesos/MesosTaskFactory.java [hunk 8]
```diff
@@ -214,8 +217,19 @@ public TaskInfo createFrom(IAssignedTask task, Offer offer) throws SchedulerExce
...
+
+            KillPolicy.Builder killPolicyBuilder = KillPolicy.newBuilder()
```
Comment (erdody) [inline src/main/java/org/apache/aurora/scheduler/mesos/MesosTaskFactory.java:229]: build() here (create the KillPolicy) and then assign the end result?
Inlining would work too.
Creating a builder is not very common.

**Pull request closed without merging**